### PR TITLE
arch/arm/common: Unify a function to restore context of tcb for context switching

### PR DESCRIPTION
**Pull request by jeongarmy**
- Unify same operations for context switching in each functions.
- It can help to avoid missing and you don't need to modify all files when some operations or values are changed.

### CODE REVIEW

File: os/arch/arm/src/common/up_restorecontext.c [hunk 1]
```diff
@@ -0,0 +1,158 @@
...
+ *
+ *   This function is called only from the TinyAra scheduling
+ *   logic.
+ *
+ * Inputs:
+ *   tcb: Refers to the tcb which will be scheduled.
+ *     This tcb is at the head of the g_readytorun task lists.
+ *
+ ************************************************************************************/
+void up_restorecontext(struct tcb_s *tcb)
```
Comment (jeongchanKim) [inline os/arch/arm/src/common/up_restorecontext.c:115]: How about unifying the up_svcall also?
There are almost same codes in up_svcall.

Comment (jeongchanKim) [reply os/arch/arm/src/common/up_restorecontext.c:115]: We already have `up_fullcontextrestore()`.
So I think the function name can be confused.

File: os/arch/arm/src/common/up_restorecontext.c [hunk 2]
```diff
@@ -0,0 +1,158 @@
...
+#include <stdint.h>
+#endif
+#ifdef CONFIG_TASK_SCHED_HISTORY
+#include <stdbool.h>
+#include <tinyara/debug/sysdbg.h>
+#endif
+#ifdef CONFIG_ARMV8M_TRUSTZONE
+#include <tinyara/tz_context.h>
+#endif
+#if defined(CONFIG_ARMV7M_MPU) || defined(CONFIG_ARMV8M_MPU)
```
Comment (jeongchanKim) [inline os/arch/arm/src/common/up_restorecontext.c:69]: How about replacing to check `CONFIG_ARM_MPU` instead of checking both?

File: os/arch/arm/src/common/up_restorecontext.c [hunk 3]
```diff
@@ -0,0 +1,158 @@
...
+{
+	/* Restore the exception context of the tcb. */
+
+#ifdef CONFIG_TASK_SCHED_HISTORY
+	/* Save the task name which will be scheduled */
+	save_task_scheduling_status(tcb);
+#endif
+
+	/* Restore the MPU registers in case we are switching to an application task */
+#if defined(CONFIG_ARMV7M_MPU) || defined(CONFIG_ARMV8M_MPU)
```
Comment (jeongchanKim) [inline os/arch/arm/src/common/up_restorecontext.c:125]: Same as above. `CONFIG_ARM_MPU`.

File: os/arch/arm/src/common/up_restorecontext.c (hunk 1 above)
Comment (Taejun-Kwon) [reply os/arch/arm/src/common/up_restorecontext.c:115]: null check for tcb should be here, even if all caller already check it to apply c secure coding style.

File: os/arch/arm/src/common/up_restorecontext.c [hunk 4]
```diff
@@ -0,0 +1,158 @@
...
+#include <tinyara/tz_context.h>
+#endif
+#if defined(CONFIG_ARMV7M_MPU) || defined(CONFIG_ARMV8M_MPU)
+#include <tinyara/mpu.h>
+#include "mpu.h"
+#endif
+#include <tinyara/arch.h>
+
+#ifdef CONFIG_TASK_MONITOR
+#include "task_monitor/task_monitor_internal.h"
```
Comment (sunghan-chang) [inline os/arch/arm/src/common/up_restorecontext.c:76]: This looks unnecessary. Could you check?

File: os/arch/arm/src/armv7-r/arm_blocktask.c [hunk 5]
```diff
@@ -143,13 +143,9 @@ void up_block_task(struct tcb_s *tcb, tstate_t task_state)
...
 			rtcb = this_task();
 
-#ifdef CONFIG_TASK_SCHED_HISTORY
-			/* Save the task name which will be scheduled */
-			save_task_scheduling_status(rtcb);
-#endif
+			/* Switch context of rtcb */
 
-			/* Then switch contexts. */
-			up_restorestate(rtcb->xcp.regs);
```
Comment (jeongchanKim) [inline os/arch/arm/src/armv7-r/arm_blocktask.c:152]: There are many similar cases in armv7-r code like arm_schedyield.c
Please check this.

File: os/arch/arm/src/common/up_restorecontext.c (hunk 1 above)
Comment (jeongarmy) [reply os/arch/arm/src/common/up_restorecontext.c:111]: @Taejun-Kwon Right, I modified. Thank you

File: os/arch/arm/src/common/up_restorecontext.c (hunk 2 above)
Comment (jeongarmy) [reply os/arch/arm/src/common/up_restorecontext.c:69]: I modified. Thank you

File: os/arch/arm/src/common/up_restorecontext.c (hunk 4 above)
Comment (jeongarmy) [reply os/arch/arm/src/common/up_restorecontext.c]: I removed. Thank you

File: os/arch/arm/src/armv7-r/arm_blocktask.c (hunk 5 above)
Comment (jeongarmy) [reply os/arch/arm/src/armv7-r/arm_blocktask.c:152]: I added missing cases. Please check again.

File: os/arch/arm/src/common/up_restoretask.c [hunk 6]
```diff
@@ -0,0 +1,152 @@
...
+	if (tcb) {
+#ifdef CONFIG_TASK_SCHED_HISTORY
+		/* Save the task name which will be scheduled */
+		save_task_scheduling_status(tcb);
+#endif
+
+		/* Restore the MPU registers in case we are switching to an application task */
+#ifdef CONFIG_ARM_MPU
+		/* Condition check : Update MPU registers only if this is not a kernel thread. */
+		if ((tcb->flags & TCB_FLAG_TTYPE_MASK) != TCB_FLAG_TTYPE_KERNEL) {
```
Comment (jeongchanKim) [inline os/arch/arm/src/common/up_restoretask.c:124]: I think this condition checking is not needed when `CONFIG_APP_BINARY_SEPARATION` is not enabled.
If we rebind the condition, we can reduce the time.

File: os/arch/arm/src/common/up_restoretask.c [hunk 7]
```diff
@@ -0,0 +1,152 @@
...
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
+ * either express or implied. See the License for the specific
+ * language governing permissions and limitations under the License.
+ *
+ ****************************************************************************/
+/****************************************************************************
+ *  arch/arm/src/common/up_restoretask.c
+ *
+ *   Copyright (C) 2007-2009 Gregory Nutt. All rights reserved.
```
Comment (jeongchanKim) [inline os/arch/arm/src/common/up_restoretask.c:21]: Is Nuttx license needed?

File: os/arch/arm/src/common/up_restoretask.c [hunk 8]
```diff
@@ -0,0 +1,105 @@
...
+		save_task_scheduling_status(tcb);
+#endif
+
+		/* Restore the MPU registers in case we are switching to an application task */
+#ifdef CONFIG_ARM_MPU
+#ifdef CONFIG_APP_BINARY_SEPARATION
+		/* Condition check : Update MPU registers only if this is not a kernel thread. */
+
+		if ((tcb->flags & TCB_FLAG_TTYPE_MASK) != TCB_FLAG_TTYPE_KERNEL) {
+			for (int i = 0; i < 3 * MPU_NUM_REGIONS; i += 3) {
```
Comment (kishore-sn) [inline os/arch/arm/src/common/up_restoretask.c:79]: Pls change the hard code value 3 to MPU_REG_NUMBER.

File: os/arch/arm/src/common/up_restoretask.c [hunk 9]
```diff
@@ -0,0 +1,105 @@
...
+ *   logic.
+ *
+ * Inputs:
+ *   tcb: Refers to the tcb which will be scheduled.
+ *     This tcb is at the head of the g_readytorun task lists.
+ *
+ ************************************************************************************/
+void up_restoretask(struct tcb_s *tcb)
+{
+	/* Restore the exception context of the tcb. */
```
Comment (kishore-sn) [inline os/arch/arm/src/common/up_restoretask.c:65]: I think this logic is not related to exception context. So we can remove this comment.

File: os/arch/arm/src/common/up_restoretask.c [hunk 10]
```diff
@@ -0,0 +1,103 @@
...
+				up_mpu_set_register(&tcb->mpu_regs[i]);
+			}
+		}
+#endif
+#ifdef CONFIG_MPU_STACK_OVERFLOW_PROTECTION
+		up_mpu_set_register(tcb->stack_mpu_regs);
+#endif
+#endif
+
+#ifdef CONFIG_SUPPORT_COMMON_BINARY
```
Comment (jeongchanKim) [inline os/arch/arm/src/common/up_restoretask.c:87]: How about moving line from 87 to line 91 within `CONFIG_APP_BINARY_SEPARATION`?
`CONFIG_SUPPORT_COMMON_BINARY` only works when `CONFIG_APP_BINARY_SEPARATION` is enabled.

File: os/arch/arm/src/armv8-m/up_unblocktask_withoutsavereg.c [hunk 11]
```diff
@@ -153,43 +144,12 @@ void up_unblock_task_without_savereg(struct tcb_s *tcb)
...
-			TZ_LoadContext_S(rtcb->tz_context);
-		}
-#endif
+		/* Restore rtcb data for context switching */
+		
+		up_restoretask(rtcb);
+		
 		/* Then switch contexts */
-
+
```
Comment (sunghan-chang) [inline os/arch/arm/src/armv8-m/up_unblocktask_withoutsavereg.c:152]: Please remove those two tabs.

File: os/arch/arm/src/common/up_restoretask.c [hunk 12]
```diff
@@ -0,0 +1,104 @@
...
+ *
+ *   This function is called only from the TinyAra scheduling
+ *   logic.
+ *
+ * Inputs:
+ *   tcb: Refers to the tcb which will be scheduled.
+ *     This tcb is at the head of the g_readytorun task lists.
+ *
+ ************************************************************************************/
+void up_restoretask(struct tcb_s *tcb)
```
Comment (sunghan-chang) [inline os/arch/arm/src/common/up_restoretask.c:63]: How about `up_restorethreadcontext`?

File: os/arch/arm/src/common/up_restoretask.c [hunk 13]
```diff
@@ -0,0 +1,104 @@
...
+/****************************************************************************
+ * Included Files
+ ****************************************************************************/
+#include <tinyara/config.h>
+#include <sched.h>
+#ifdef CONFIG_SUPPORT_COMMON_BINARY
+#include <stdint.h>
+#endif
+#ifdef CONFIG_TASK_SCHED_HISTORY
+#include <stdbool.h>
```
Comment (sunghan-chang) [inline os/arch/arm/src/common/up_restoretask.c:27]: This looks unnecessary.

File: os/arch/arm/src/common/up_restoretask.c (hunk 12 above)
Comment (jeongarmy) [reply os/arch/arm/src/common/up_restoretask.c:63]: I removed 'context' because the functions with 'context' like up_switchcontext, up_fullcontextrestore switches registers but this just sets tcb data needed for context switching.

File: os/arch/arm/src/common/up_restoretask.c (hunk 13 above)
Comment (jeongarmy) [reply os/arch/arm/src/common/up_restoretask.c:27]: It is used for CONFIG_TASK_MONITOR so I modified.

File: os/arch/arm/src/armv7-m/up_schedyield.c [hunk 14]
```diff
@@ -144,41 +126,14 @@ void up_schedyield(void)
...
-			ntcb->is_active = true;
-#endif
-
+			/* Restore ntcb data for context switching */
+			
+			up_restoretask(ntcb);
+			
 			/* Then switch contexts */
-
+
```
Comment (sunghan-chang) [inline os/arch/arm/src/armv7-m/up_schedyield.c:134]: Please remove them.

File: os/include/tinyara/arch.h [hunk 15]
```diff
@@ -432,6 +432,23 @@ void up_allocate_secure_context(TZ_ModuleId_t size);
 void up_free_secure_context();
 #endif
 
+/************************************************************************************
+ * Name: up_restoretask
+ *
+ * Description:
+ *   Restore and set data needed for context switching from tcb then switches context.
```
Comment (jeongchanKim) [inline os/include/tinyara/arch.h:439]: This function doesn't make context switching.
Please check and modify the description.

**Review by sunghan-chang — APPROVED** [review]

File: os/arch/arm/src/common/up_restorecontext.c (hunk 3 above)
Comment (jeongarmy) [reply os/arch/arm/src/common/up_restorecontext.c]: I modified
